README.md
# React Calendar Timeline

This is new version of the [react-calendar-timeline](https://github.com/namespace-ee/react-calendar-timeline). All open pull requests were merged.

A modern and responsive react timeline component.

![calendar demo](https://raw.githubusercontent.com/namespace-ee/react-calendar-timeline/master/demo.gif)

Demo here: http://namespace.ee/react-calendar-timeline/

## Getting started

```bash
# via yarn
yarn add new-react-calendar-timeline

# via npm
npm install --save new-react-calendar-timeline
```

`new-react-calendar-timeline` has `react`, `react-dom`, [`moment`](http://momentjs.com/) and [`interact.js`](http://interactjs.io/docs/) as peer dependencies.

You need to install them separately:

```bash
# via yarn
yarn add react react-dom # you probably already have these
yarn add moment interact.js

# via npm
npm install --save react react-dom # you probably already have these
npm install --save moment interact.js
```

## Usage

At the very minimum:

```jsx
import Timeline from 'react-calendar-timeline'
import moment from 'moment'

const groups = [
  {id: 1, title: 'group 1'},
  {id: 2, title: 'group 2'}
]

const items = [
  {id: 1, group: 1, title: 'item 1', start_time: moment(), end_time: moment().add(1, 'hour')},
  {id: 2, group: 2, title: 'item 2', start_time: moment().add(-0.5, 'hour'), end_time: moment().add(0.5, 'hour')},
  {id: 3, group: 1, title: 'item 3', start_time: moment().add(2, 'hour'), end_time: moment().add(3, 'hour')}
]

ReactDOM.render(
  <div>
    Rendered by react!
    <Timeline groups={groups}
              items={items}
              defaultTimeStart={moment().add(-12, 'hour')}
              defaultTimeEnd={moment().add(12, 'hour')}
              />
  </div>,
  document.getElementById('root')
);
```

## API
*NB!* All props need to be immutable. For example, this means if you wish to change the title of one of your items, please pass in a whole new items array instead of changing the title in the old array. [Here's more info.](http://reactkungfu.com/2015/08/pros-and-cons-of-using-immutability-with-react-js/)

The component can take many props:

### groups
Expects either a vanilla JS array or an immutableJS array, consisting of objects with the following attributes:
```js
{
  id: 1,
  title: 'group 1',
  rightTitle: 'title in the right sidebar'
}
```

If you use right sidebar, you can pass optional `rightTitle` property here.

### items
Expects either a vanilla JS array or an immutableJS array, consisting of objects with the following attributes:
```js
{
  id: 1,
  group: 1,
  title: 'Random title',
  start_time: 1457902922261,
  end_time: 1457902922261 + 86400000,
  canMove: true,
  canResize: false,
  canChangeGroup: false,
  className: 'weekend',
  itemProps: {
    // these optional attributes are passed to the root <div /> of each item as <div {...itemProps} />
    'data-custom-attribute': 'Random content',
    'aria-hidden': true,
    onDoubleClick: () => { console.log('You clicked double!') }
  }
}
```

The preferred (fastest) option is to give unix timestamps in milliseconds for `start_time` and `end_time`. Objects that convert to them (JavaScript Date or moment()) will also work, but will be a lot slower.

### selected
An array with id's corresponding to id's in items (`item.id`). If this prop is set you have to manage the selected items yourself within the `onItemSelect` handler to update the property with new id's. This overwrites the default behaviour of selecting one item on click.

### keys
An array specifying keys in the `items` and `groups` objects. Defaults to
```js
{
  groupIdKey: 'id',
  groupTitleKey: 'title',
  groupRightTitleKey: 'rightTitle',
  itemIdKey: 'id',
  itemTitleKey: 'title',    // key for item div content
  itemDivTitleKey: 'title', // key for item div title (<div title="text"/>)
  itemGroupKey: 'group',
  itemTimeStartKey: 'start_time',
  itemTimeEndKey: 'end_time'
}
```

### sidebarWidth
Width of the sidebar in pixels. If set to `0`, the sidebar is not rendered. Defaults to `150`.

### sidebarContent
Everything passed here will be displayed above the left sidebar. Use this to display small filters or so. Defaults to `null`.

### rightSidebarWidth
Width of the right sidebar in pixels. If set to `0`, the right sidebar is not rendered. Defaults to `0`.

### rightSidebarContent
Everything passed here will be displayed above the right sidebar. Use this to display small filters or so. Defaults to `null`.

### dragSnap
Snapping unit when dragging items. Defaults to `15 * 60 * 1000` or 15min. When so, the items will snap to 15min intervals when dragging.

### minResizeWidth
The minimum width, in pixels, of a timeline entry when it's possible to resize. If not reached, you must zoom in to resize more. Default to `20`.

### fixedHeader
How does the header (the scrolling part with dates) behave if not all of the groups fit on the page, resulting in a vertical scrollbar.

* `fixed` - the header is always fixed to its initial position
* `sticky` (default) - the header follows the scroll of the page to be always visible
* `none` - the header is always at the top of the component and doesn't stick with scrolling

### stickyOffset
If fixedHeader is `sticky`, at what height from the top of the screen should we start floating it? This is useful if for example you already have
a sticky navbar. Defaults `0`.

### fullUpdate
If your calendar has large items compared to the zoom level (e.g. multi week events when viewing one day at a time), set this to `true` (default).

If you have many small events compared to the zoom level (e.g. hundreds of 30min events and viewing one week at a time), set this to `false`.

When set to `true` we update the dimensions of the items on every scroll event. This looks nicer, as 1) item labels
are always fully on the screen, even if the start or end of the items is off screen, 2) item stacking also reflects what's on the screen.

When set to `false`, we update the dimensions of the items only when the [scrolling canvas](https://github.com/namespace-ee/react-calendar-timeline#behind-the-scenes) updates. This makes scrolling much faster, but labels can go off screen.

### lineHeight
Height of one line in the calendar in pixels. Default `30`

### headerLabelGroupHeight
Height of the top header line. Default `30`

### headerLabelHeight
Height of the bottom header line. Default `30`

### itemHeightRatio
What percentage of the height of the line is taken by the item? Default `0.65`

### minZoom
Smallest time the calendar can zoom to in milliseconds. Default `60 * 60 * 1000` (1 hour)

### maxZoom
Largest time the calendar can zoom to in milliseconds. Default `5 * 365.24 * 86400 * 1000` (5 years)

### clickTolerance
How many pixels we can drag the background for it to be counted as a click on the background. Defualt: `3`

### canMove
Can items be dragged around? Can be overridden in the `items` array. Defaults to `true`

### canChangeGroup
Can items be moved between groups? Can be overridden in the `items` array. Defaults to `true`

### canResize
Can items be resized? Can be overridden in the `items` array. Accepted values: `false`, `"left"`, `"right"`, `"both"`. Defaults to `"right"`. If you pass `true`, it will be treated as `"right"` to not break compatibility with versions 0.9 and below.

### useResizeHandle
Append a special `.rct-drag-right` handle to the elements and only resize if dragged from there. Defaults to `false`

### showCursorLine
Show a vertical line at the snap point when you mouse over the calendar

### stackItems
Stack items under each other, so there is no visual overlap when times collide. Defaults to `false`.

### traditionalZoom
Zoom in when scrolling the mouse up/down. Defaults to `false`

### itemTouchSendsClick
Normally tapping (touching) an item selects it. If this is set to true, a tap will have the same effect, as selecting with the first click and then clicking again to open and send the onItemClick event. Defaults to `false`.

### timeSteps
With what step to display different units. E.g. `15` for `minute` means only minutes 0, 15, 30 and 45 will be shown.

Default:
```js
{
  second: 1,
  minute: 1,
  hour: 1,
  day: 1,
  month: 1,
  year: 1
}
```

### onItemMove(itemId, dragTime, newGroupOrder)
Callback when an item is moved. Returns 1) the item's ID, 2) the new start time and 3) the index of the new group in the `groups` array.

### onItemResize(itemId, time, edge)
Callback when an item is resized. Returns 1) the item's ID, 2) the new start or end time of the item 3) The edge that was dragged (`left` or `right`)

### onItemSelect(itemId, e)
Called when an item is selected. This is sent on the first click on an item.

### onItemClick(itemId, e)
Called when an item is clicked. Note: the item must be selected before it's clicked... except if it's a touch event and `itemTouchSendsClick` is enabled.

### onItemDoubleClick(itemId, e)
Called when an item was double clicked

### onItemContextMenu(itemId, e)
Called when the item is clicked by the right button of the mouse. Note: If this property is set the default context menu doesn't appear

### onCanvasClick(groupId, time, e)
Called when an empty spot on the canvas was clicked. Get the group ID and the time as arguments. For example open a "new item" window after this.

### onCanvasDoubleClick(groupId, time, e)
Called when an empty spot on the canvas was double clicked. Get the group ID and the time as arguments.

### onCanvasContextMenu(group, time, e)
Called when the canvas is clicked by the right button of the mouse. Note: If this property is set the default context menu doesn't appear

### moveResizeValidator(action, itemId, time, resizeEdge)
This function is called when an item is being moved or resized. It's up to this function to return a new version of `change`, when the proposed move would violate business logic.

The argument `action` is one of `move` or `resize`.

The argument `resizeEdge` is when resizing one of `left` or `right`.

The argument `time` describes the proposed new time for either the start time of the item (for move) or the start or end time (for resize).

The function must return a new unix timestamp in milliseconds... or just `time` if the proposed new time doesn't interfere with business logic.

For example, to prevent moving of items into the past, but to keep them at 15min intervals, use this code:

```js
function (action, item, time, resizeEdge) {
  if (time < new Date().getTime()) {
    var newTime = Math.ceil(new Date().getTime() / (15*60*1000)) * (15*60*1000);
    return newTime;
  }

  return time
}
```

### moveGroupValidator(item, targetGroupOrder, groupDelta, previousGroupDelta) ###
This function is called when an item jumps to a new group when moved. It's up to this function to return a new version of `groupDelta` when the proposed group jump would violate business logic.

The argument `item` represents the object of the item that is being moved.

The argument `targetGroupOrder` is an index of the target group, a group into which the items is being moved.

The argument `groupDelta` represents the change in group index. The change is between the item's original group index and the target group index, into which the item is being moved.

The argument `previousGroupDelta` represents the change in group index from *right before the jump*. In other words, it is an indicator from where the item jumps.

For example, let's assume that our business logic is that item with ID of 5 shouldn't move to third group on our list. When the user attempts moving to group 3, we want to block this action and keep the item in its previous group. This is where the `previousGroupDelta` will be useful - we can return this value to tell the item to stay in the previous group.

```js
function (item, targetGroupOrder, groupDelta, previousGroupDelta) {
  if (item.id === 5 && targetGroupOrder === 2) {
    return previousGroupDelta;
  }

  return groupDelta;
}
```

Another example would be a situation where we don't want any item to jump more than three neighbouring groups.

```js
function (item, targetGroupOrder, groupDelta, previousGroupDelta) {
  if (groupDelta > 3 || groupDelta < -3) {
    return previousGroupDelta;
  }

  return groupDelta;
}
```

### defaultTimeStart and defaultTimeEnd
Unless overridden by `visibleTimeStart` and `visibleTimeEnd`, specify where the calendar begins and where it ends. This parameter expects a Date or moment object.

### visibleTimeStart and visibleTimeEnd
The exact viewport of the calendar. When these are specified, scrolling in the calendar must be orchestrated by the `onTimeChange` function.  This parameter expects a unix timestamp in milliseconds.

### headerLabelFormats and subHeaderLabelFormats
The formats passed to moment to render times in the header and subheader. Defaults to these:

```js
import { defaultHeaderLabelFormats, defaultSubHeaderLabelFormats } from 'react-calendar-timeline'

defaultHeaderLabelFormats == {
  yearShort: 'YY',
  yearLong: 'YYYY',
  monthShort: 'MM/YY',
  monthMedium: 'MM/YYYY',
  monthMediumLong: 'MMM YYYY',
  monthLong: 'MMMM YYYY',
  dayShort: 'L',
  dayLong: 'dddd, LL',
  hourShort: 'HH',
  hourMedium: 'HH:00',
  hourMediumLong: 'L, HH:00',
  hourLong: 'dddd, LL, HH:00',
  time: 'LLL'
}

defaultSubHeaderLabelFormats == {
  yearShort: 'YY',
  yearLong: 'YYYY',
  monthShort: 'MM',
  monthMedium: 'MMM',
  monthLong: 'MMMM',
  dayShort: 'D',
  dayMedium: 'dd D',
  dayMediumLong: 'ddd, Do',
  dayLong: 'dddd, Do',
  hourShort: 'HH',
  hourLong: 'HH:00',
  minuteShort: 'mm',
  minuteLong: 'HH:mm'
}
```

For US time formats (AM/PM), use these:

```js
import { defaultHeaderLabelFormats, defaultSubHeaderLabelFormats } from 'react-calendar-timeline'

const usHeaderLabelFormats = Object.assign({}, defaultSubHeaderLabelFormats, {
  hourShort: 'h A',
  hourMedium: 'h A',
  hourMediumLong: 'L, h A',
  hourLong: 'dddd, LL, h A',
})

const usSubHeaderLabelFormats = Object.assign({}, defaultSubHeaderLabelFormats, {
  hourShort: 'h A',
  hourLong: 'h A',
  minuteLong: 'h:mm A'
})
```

... and then pass these as `headerLabelFormats` and `subHeaderLabelFormats`

### onTimeChange(visibleTimeStart, visibleTimeEnd, updateScrollCanvas)
A function that's called when the user tries to scroll. Call the passed `updateScrollCanvas(start, end)` with the updated visibleTimeStart and visibleTimeEnd (as unix timestamps in milliseconds) to change the scroll behavior, for example to limit scrolling.

Here is an example that limits the timeline to only show dates starting 6 months from now and ending in 6 months.

```js

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres (more or less) to [Semantic Versioning](http://semver.org/).

## Unreleased
When you submit a PR, add your changes here!

## [2.2.0]
### New version
- Merged with source v0.15.1

## [2.0.0]
### New version
- Merged all open pull request and create new version of the plugin

## [0.15.0]
Plugin support and sticky header!

### Added
- Plugins system (pass them as children) @mariusandra #122
- Sticky header (`fixedHeader='sticky'`) @mariusandra #125

### Removed
- [BREAKING] Removed deprecated option to pass sidebar header content as children. Use `sidebarContent` instead. @mariusandra
- [BREAKING] Removed fixedHeader option `absolute`, which was broken and is now replaced with the option `sticky` @mariusandra

### Demo & Docs
- Notice for modern module bundlers @jlubben @mariusandra #128
- Add [treeGroups](http://namespace.ee/react-calendar-timeline-docs/#/treeGroups) demo

## [0.14.11]
Plenty of bugfixes, tests and new demos in these 0.14 patch releases.

### Fixed
- Fixed bug with `resizeDetector` and with detecting changes in `sidebarWidth` @mariusandra
- Fixed bug where order `0` was evaluated as a falsy @nicocrm #111
- Fix overflow-x with header @signalwerk

### Added
- Add meta+wheel modifier that zooms 3x the speed of the normal wheel events @mariusandra

### Changed
- Refactor `calculateDimensions` to be pure @signalwerk
- Convert `groupHeights` and `groupTops` to arrays (from objects) @mariusandra

### Demo & docs
- Add [linkedTimelines](http://namespace.ee/react-calendar-timeline-docs/#/linkedTimelines) demo
- Add [elementResize](http://namespace.ee/react-calendar-timeline-docs/#/elementResize) demo
- Add docs about modifier keys for zooming/scrolling @signalwerk

## [0.14.2]
### Changed
- Use `prop-types` instead of `React.PropTypes` to support React 15.5+. @mariusandra #110

## [0.14.0]
### Added
- Use `headerLabelFormats` and `subHeaderLabelFormats` to customise the header labels. @Slowyn #68
- Optional pluggable `resizeDetector` to detect when the element's container is resized. @Ziller321 #94

### Fixed
- Fix renders with empty `groups` array. @signalwerk #106

## [0.13.0]
### Added
- An option to add another sidebar to the right of the Timeline. @goooseman #80
- `itemRenderer` prop to allow specifying a custom component to render the items @nicocrm #103
- `groupRenderer` prop to allow specifying a custom component to render the groups @nicocrm #103
- `showCursorLine` prop to show a vertical line at the snap position @meikoudras
- You can now select multiple items if you take control of the `selected` prop and the `onItemSelect` handler. @meengit #71
- Canvas context menu handler `onCanvasContextMenu` @meikoudras

### Fixed
- Calculate width when we receive sidebar width property @jmerriweather #75
- Avoid updating updateDimensions right after updateScrollCanvas @nicocrm #87
- Fix typo collision detection in stack() @nicocrm #96
- Remove dead code @signalwerk #101
- Disable cursor style by interactjs @bkniffler #89
- Fixed header width and Header label weekday support @meikoudras #66

### Changed
- [Deprecated] To have content above the left sidebar, pass it in a `sidebarContent={<div />}` prop, not as children to the component.

## [0.11.1]
### Fixed
- Without canResize prop in items it gave a Uncaught TypeError. @tgosp

## [0.11.0]
### Added
- An option to fully update the calendar at every scroll event. With this change, labels of items are always fully visible, even if looking at a multi day event with a zoom level set at 30min. @mariusandra

## [0.10.1]
### Changed
- The left resize edge mouse cursor is now a left arrow @mariusandra

## [0.10.0]
### Added
- You can also resize items from the left now @mariusandra

## [0.9.0]
### Added
- Allow disabling selection clicks on items #58 by @sjchmiela
- Allow passing additional props to `Item`'s `<div/>` #58 by @sjchmiela
- Add `clickTolerance` so dragging more than 3 pixels is no longer a click @mariusandra

### Changed
- [BREAKING] Same arguments order (groupId, time, e) for onCanvasDoubleClick and onCanvasClick #52 by @signalwerk
- [Deprecated] `onTimeChange` now gets `updateScrollCanvas` as the third argument. Doing `this.updateScrollCanvas` is no longer needed and will be removed soon.
- Moved React & Moment from dependencies to peerDependencies #53 by @meikoudras
- Fix resizing when inside DIV #47 by @semargal
- Fix demo for IE11 #44 by @lucidlemon
- Package a .css file, not a .scss file as previously done. @mariusandra

[0.9.0]: https://github.com/namespace-ee/react-calendar-timeline/compare/v0.8.6...v0.9.0
[0.10.0]: https://github.com/namespace-ee/react-calendar-timeline/compare/v0.9.0...v0.10.0
[0.10.1]: https://github.com/namespace-ee/react-calendar-timeline/compare/v0.10.0...v0.10.1
[0.11.0]: https://github.com/namespace-ee/react-calendar-timeline/compare/v0.10.1...v0.11.0
[0.11.1]: https://github.com/namespace-ee/react-calendar-timeline/compare/v0.11.0...v0.11.1
[0.13.0]: https://github.com/namespace-ee/react-calendar-timeline/compare/v0.11.1...v0.13.0
[0.14.0]: https://github.com/namespace-ee/react-calendar-timeline/compare/v0.13.0...v0.14.0
[0.14.2]: https://github.com/namespace-ee/react-calendar-timeline/compare/v0.14.0...v0.14.2
[0.14.11]: https://github.com/namespace-ee/react-calendar-timeline/compare/v0.14.2...v0.14.11
[0.15.0]: https://github.com/namespace-ee/react-calendar-timeline/compare/v0.14.11...v0.15.0


src/lib/items/Item.js
import React, { Component } from 'react'
import interact from 'interact.js'
import moment from 'moment'

import { _get, deepObjectCompare } from '../utils'

export default class Item extends Component {
  // removed prop type check for SPEED!
  // they are coming from a trusted component anyway
  // (this complicates performance debugging otherwise)
  static propTypes = {
    // canvasTimeStart: React.PropTypes.number.isRequired,
    // canvasTimeEnd: React.PropTypes.number.isRequired,
    // canvasWidth: React.PropTypes.number.isRequired,
    // order: React.PropTypes.number.isRequired,
    //
    // dragSnap: React.PropTypes.number,
    // minResizeWidth: React.PropTypes.number,
    // selected: React.PropTypes.bool,
    //
    // canChangeGroup: React.PropTypes.bool.isRequired,
    // canMove: React.PropTypes.bool.isRequired,
    // canResizeLeft: React.PropTypes.bool.isRequired,
    // canResizeRight: React.PropTypes.bool.isRequired,
    //
    // keys: React.PropTypes.object.isRequired,
    // item: React.PropTypes.object.isRequired,
    //
    // onSelect: React.PropTypes.func,
    // onDrag: React.PropTypes.func,
    // onDrop: React.PropTypes.func,
    // onResizing: React.PropTypes.func,
    // onResized: React.PropTypes.func,
    // onContextMenu: React.PropTypes.func,
    // itemRenderer: React.PropTypes.func
  }

  static defaultProps = {
    selected: false
  }

  constructor (props) {
    super(props)

    this.cacheDataFromProps(props)

    this.state = {
      interactMounted: false,

      dragging: null,
      dragStart: null,
      preDragPosition: null,
      dragTime: null,
      dragGroupDelta: null,
      targetDragGroupDelta: null,

      resizing: null,
      resizeEdge: null,
      resizeStart: null,
      resizeTime: null
    }
  }

  shouldComponentUpdate (nextProps, nextState) {
    var shouldUpdate = nextState.dragging !== this.state.dragging ||
                       nextState.dragTime !== this.state.dragTime ||
                       nextState.dragGroupDelta !== this.state.dragGroupDelta ||
                       nextState.resizing !== this.state.resizing ||
                       nextState.resizeTime !== this.state.resizeTime ||
                       nextProps.keys !== this.props.keys ||
                       !deepObjectCompare(nextProps.itemProps, this.props.itemProps) ||
                       nextProps.selected !== this.props.selected ||
                       nextProps.item !== this.props.item ||
                       nextProps.canvasTimeStart !== this.props.canvasTimeStart ||
                       nextProps.canvasTimeEnd !== this.props.canvasTimeEnd ||
                       nextProps.canvasWidth !== this.props.canvasWidth ||
                       nextProps.order !== this.props.order ||
                       nextProps.dragSnap !== this.props.dragSnap ||
                       nextProps.minResizeWidth !== this.props.minResizeWidth ||
                       nextProps.canChangeGroup !== this.props.canChangeGroup ||
                       nextProps.canSelect !== this.props.canSelect ||
                       nextProps.topOffset !== this.props.topOffset ||
                       nextProps.canMove !== this.props.canMove ||
                       nextProps.canResizeLeft !== this.props.canResizeLeft ||
                       nextProps.canResizeRight !== this.props.canResizeRight ||
                       nextProps.dimensions !== this.props.dimensions
    return shouldUpdate
  }

  cacheDataFromProps (props) {
    this.itemId = _get(props.item, props.keys.itemIdKey)
    this.itemTitle = _get(props.item, props.keys.itemTitleKey)
    this.itemDivTitle = props.keys.itemDivTitleKey ? _get(props.item, props.keys.itemDivTitleKey) : this.itemTitle
    this.itemTimeStart = _get(props.item, props.keys.itemTimeStartKey)
    this.itemTimeEnd = _get(props.item, props.keys.itemTimeEndKey)
  }

  coordinateToTimeRatio (props = this.props) {
    return (props.canvasTimeEnd - props.canvasTimeStart) / props.canvasWidth
  }

  dragTimeSnap (dragTime, considerOffset) {
    const { dragSnap } = this.props
    if (dragSnap) {
      const offset = considerOffset ? moment().utcOffset() * 60 * 1000 : 0
      return Math.round(dragTime / dragSnap) * dragSnap - offset % dragSnap
    } else {
      return dragTime
    }
  }

  resizeTimeSnap (dragTime) {
    const { dragSnap } = this.props
    if (dragSnap) {
      const endTime = this.itemTimeEnd % dragSnap
      return Math.round((dragTime - endTime) / dragSnap) * dragSnap + endTime
    } else {
      return dragTime
    }
  }

  dragTime (e) {
    const startTime = this.itemTimeStart

    if (this.state.dragging) {
      const deltaX = e.pageX - this.state.dragStart.x
      const timeDelta = deltaX * this.coordinateToTimeRatio()

      return this.dragTimeSnap(startTime + timeDelta, true)
    } else {
      return startTime
    }
  }

  dragGroupDelta (e) {
    const {groupTops, order, topOffset} = this.props
    if (this.state.dragging) {
      if (!this.props.canChangeGroup) {
        return 0
      }
      let groupDelta = 0

      for (var key of Object.keys(groupTops)) {
        var item = groupTops[key]
        if (e.pageY - topOffset > item) {
          groupDelta = parseInt(key, 10) - order
        } else {
          break
        }
      }

      if (this.props.order + groupDelta < 0) {
        return 0 - this.props.order
      } else {
        return groupDelta
      }
    } else {
      return 0
    }
  }

  resizeTimeDelta (e, resizeEdge) {
    const length = this.itemTimeEnd - this.itemTimeStart
    const timeDelta = this.dragTimeSnap((e.pageX - this.state.resizeStart) * this.coordinateToTimeRatio())

    if (length + (resizeEdge === 'left' ? -timeDelta : timeDelta) < (this.props.dragSnap || 1000)) {
      if (resizeEdge === 'left') {
        return length - (this.props.dragSnap || 1000)
      } else {
        return (this.props.dragSnap || 1000) - length
      }
    } else {
      return timeDelta
    }
  }

  componentDidMount () {
  }

  mountInteract () {
    const leftResize = this.props.useResizeHandle ? this.refs.dragLeft : true
    const rightResize = this.props.useResizeHandle ? this.refs.dragRight : true

    interact(this.refs.item)
      .resizable({
        edges: {
          left: this.canResizeLeft() && leftResize,
          right: this.canResizeRight() && rightResize,
          top: false,
          bottom: false
        },
        enabled: this.props.selected && (this.canResizeLeft() || this.canResizeRight())
      })
      .draggable({
        enabled: this.props.selected
      })
      .styleCursor(false)
      .on('dragstart', (e) => {
        if (this.props.selected) {
          this.setState({
            dragging: true,
            dragStart: {x: e.pageX, y: e.pageY},
            preDragPosition: {x: e.target.offsetLeft, y: e.target.offsetTop},
            dragTime: this.itemTimeStart,
            dragGroupDelta: 0
          })
        } else {
          return false
        }
      })
      .on('dragmove', (e) => {
        if (this.state.dragging) {
          let dragTime = this.dragTime(e)
          let dragGroupDelta = this.dragGroupDelta(e)

          if (this.props.moveGroupValidator) {
            dragGroupDelta = this.props.moveGroupValidator(
              this.props.item,
              this.props.order + dragGroupDelta,
              dragGroupDelta,
              this.state.dragGroupDelta,
              dragTime
            )
          }

          if (this.props.moveResizeValidator) {
            dragTime = this.props.moveResizeValidator(
              'move',
              this.props.item,
              dragTime,
              this.props.order + dragGroupDelta,
              this.state.dragTime
            )
          }

          if (this.props.onDrag) {
            this.props.onDrag(this.itemId, dragTime, this.props.order + dragGroupDelta)
          }

          this.setState({
            dragTime: dragTime,
            dragGroupDelta: dragGroupDelta
          })
        }
      })
      .on('dragend', (e) => {
        if (this.state.dragging) {
          if (this.props.onDrop) {
            this.props.onDrop(this.itemId, this.state.dragTime, this.props.order + this.state.dragGroupDelta)
          }

          this.setState({
            dragging: false,
            dragStart: null,
            preDragPosition: null,
            dragTime: null,
            dragGroupDelta: null
          })
        }
      })
      .on('resizestart', (e) => {
        if (this.props.selected) {
          this.setState({
            resizing: true,
            resizeEdge: null, // we don't know yet
            resizeStart: e.pageX,
            resizeTime: 0
          })
        } else {
          return false
        }
      })
      .on('resizemove', (e) => {
        if (this.state.resizing) {
          let resizeEdge = this.state.resizeEdge

          if (!resizeEdge) {
            resizeEdge = e.deltaRect.left !== 0 ? 'left' : 'right'
            this.setState({ resizeEdge })
          }
          const time = resizeEdge === 'left' ? this.itemTimeStart : this.itemTimeEnd

          let resizeTime = this.resizeTimeSnap(time + this.resizeTimeDelta(e, resizeEdge))

          if (this.props.moveResizeValidator) {
            resizeTime = this.props.moveResizeValidator('resize', this.props.item, resizeTime, resizeEdge, this.state.resizeTime)
          }

          if (this.props.onResizing) {
            this.props.onResizing(this.itemId, resizeTime, resizeEdge)
          }

          this.setState({
            resizeTime
          })
        }
      })
      .on('resizeend', (e) => {
        if (this.state.resizing) {
          const { resizeEdge } = this.state
          const time = resizeEdge === 'left' ? this.itemTimeStart : this.itemTimeEnd
          let resizeTime = this.resizeTimeSnap(time + this.resizeTimeDelta(e, resizeEdge))

          if (this.props.moveResizeValidator) {
            resizeTime = this.props.moveResizeValidator('resize', this.props.item, resizeTime, resizeEdge)
          }

          if (this.props.onResized && this.resizeTimeDelta(e, resizeEdge) !== 0) {
            this.props.onResized(this.itemId, resizeTime, resizeEdge)
          }
          this.setState({
            resizing: null,
            resizeStart: null,
            resizeEdge: null,
            resizeTime: null
          })
        }
      })
      .on('tap', (e) => {
        this.actualClick(e, e.pointerType === 'mouse' ? 'click' : 'touch')
      })

    this.setState({
      interactMounted: true
    })
  }

  canResizeLeft (props = this.props) {
    if (!props.canResizeLeft || props.dimensions.clippedLeft) {
      return false
    }
    let width = parseInt(props.dimensions.width, 10)
    return width >= props.minResizeWidth
  }

  canResizeRight (props = this.props) {
    if (!props.canResizeRight || props.dimensions.clippedRight) {
      return false
    }
    let width = parseInt(props.dimensions.width, 10)
    return width >= props.minResizeWidth
  }

  canMove (props = this.props) {
    return !!props.canMove
  }

  componentWillReceiveProps (nextProps) {
    this.cacheDataFromProps(nextProps)

    let { interactMounted } = this.state
    const couldDrag = this.props.selected && this.canMove(this.props)
    const couldResizeLeft = this.props.selected && this.canResizeLeft(this.props)
    const couldResizeRight = this.props.selected && this.canResizeRight(this.props)
    const willBeAbleToDrag = nextProps.selected && this.canMove(nextProps)
    const willBeAbleToResizeLeft = nextProps.selected && this.canResizeLeft(nextProps)
    const willBeAbleToResizeRight = nextProps.selected && this.canResizeRight(nextProps)

    if (nextProps.selected && !interactMounted) {
      this.mountInteract()
      interactMounted = true
    }

    if (interactMounted && (couldResizeLeft !== willBeAbleToResizeLeft || couldResizeRight !== willBeAbleToResizeRight)) {
      const leftResize = this.props.useResizeHandle ? this.refs.dragLeft : true
      const rightResize = this.props.useResizeHandle ? this.refs.dragRight : true

      interact(this.refs.item)
        .resizable({
          enabled: willBeAbleToResizeLeft || willBeAbleToResizeRight,
          edges: {
            top: false,
            bottom: false,
            left: willBeAbleToResizeLeft && leftResize,
            right: willBeAbleToResizeRight && rightResize
          }
        })
    }
    if (interactMounted && couldDrag !== willBeAbleToDrag) {
      interact(this.refs.item)
        .draggable({enabled: willBeAbleToDrag})
    }
  }

  onMouseDown = (e) => {
    if (!this.state.interactMounted) {
      e.preventDefault()
      this.startedClicking = true
    }
  };

  onMouseUp = (e) => {
    if (!this.state.interactMounted && this.startedClicking) {
      this.startedClicking = false
      this.actualClick(e, 'click')
    }
  };

  onTouchStart = (e) => {
    if (!this.state.interactMounted) {
      e.preventDefault()
      this.startedTouching = true
    }
  };

  onTouchEnd = (e) => {
    if (!this.state.interactMounted && this.startedTouching) {
      this.startedTouching = false
      this.actualClick(e, 'touch')
    }
  };

  handleDoubleClick = (e) => {
    e.preventDefault()
    e.stopPropagation()
    if (this.props.onItemDoubleClick) {
      this.props.onItemDoubleClick(this.itemId, e)
    }
  };

  handleContextMenu = (e) => {
    if (this.props.onContextMenu) {
      e.preventDefault()
      e.stopPropagation()
      this.props.onContextMenu(this.itemId, e)
    }
  };

  actualClick (e, clickType) {
    if (this.props.canSelect && this.props.onSelect) {
      this.props.onSelect(this.itemId, clickType, e)
    }
  }

  renderContent () {
    const Comp = this.props.itemRenderer
    if (Comp) {
      return <Comp item={this.props.item} />
    } else {
      return this.itemTitle
    }
  }

  render () {
    const dimensions = this.props.dimensions
    if (typeof this.props.order === 'undefined' || this.props.order === null) {
      return null
    }

    const classNames = 'rct-item' +
                       (this.props.selected ? ' selected' : '') +
                       (this.canMove(this.props) ? ' can-move' : '') +
                       (this.canResizeLeft(this.props) || this.canResizeRight(this.props) ? ' can-resize' : '') +
                       (this.canResizeLeft(this.props) ? ' can-resize-left' : '') +
                       (this.canResizeRight(this.props) ? ' can-resize-right' : '') +
                       (this.props.item.className ? ` ${this.props.item.className}` : '') +
                       (dimensions.clippedLeft ? ' clipped-left' : '') +
                       (dimensions.clippedRight ? ' clipped-right' : '')

    const style = {
      left: `${dimensions.left}px`,
      top: `${dimensions.top}px`,
      width: `${dimensions.width}px`,
      height: `${dimensions.height}px`,
      lineHeight: `${dimensions.height}px`
    }

    return (
      <div {...this.props.item.itemProps}
           key={this.itemId}
           ref='item'
           className={classNames}
           title={this.itemDivTitle}
           onMouseDown={this.onMouseDown}
           onMouseUp={this.onMouseUp}
           onTouchStart={this.onTouchStart}
           onTouchEnd={this.onTouchEnd}
           onDoubleClick={this.handleDoubleClick}
           onContextMenu={this.handleContextMenu}
           style={style}>
        {this.props.useResizeHandle ? <div ref='dragLeft' className='rct-drag-left'></div> : ''}
        <div className='rct-item-overflow'>
          <div className='rct-item-content'>
            {this.renderContent()}
          </div>
        </div>
        {this.props.useResizeHandle ? <div ref='dragRight' className='rct-drag-right'></div> : ''}
      </div>
    )
  }
}
